### PR TITLE
Let splat() specify default arguments for .f

### PR DESCRIPTION
**Pull request by lionelgit**
This lets `splat()` accept additional arguments that are fed to `.f` as defaults.

When we pipe a list into a splatted function, it is often necessary to take care of additional arguments. That typically means adjusting the main list with additional arguments by using a temporary variable or `append()`ing  them in the middle of the pipe chain. Being able to specify these arguments directly in `splat()` hopefully leads to clearer code because the main arguments and the side arguments are specified separately (the former in the pipe chain and the latter in `splat()`).

### CODE REVIEW

**Comment by lionelgit** [comment]
In #33 I mentioned the idea of having
> differing function signatures as a function of the type of the first argument

Along these lines, it could also make sense to have a dual behavior for `splat()` such it would behave as a functional or as a factory function depending on the type of the first argument.
```{r}
splat(fun)(x) # factory function
splat(x, fun) # functional (equivalent to do.call with x as first arg)
```

I don't like factory functions so much in pipes because the parentheses are syntax heavy. With the dual behavior we could write
```{r}
x %>%
  map(fun1) %>%
  splat(fun2)
```
instead of
```{r}
x %>%
  map(fun1) %>%
  splat(fun2)()
```

File: R/splat.R [hunk 1]
```diff
@@ -1,11 +1,14 @@
...
-#' This is useful when you want to pass a function a row of data frame or
-#' list, and don't want to manually pull it apart in your function.
+#' arguments, it takes a single named list which will be interpreted
+#' as its arguments. This is useful when you want to pass a function
+#' a row of data frame or list, and don't want to manually pull it
+#' apart in your function. You can also specify default arguments for
+#' \code{.f} in the call to \code{splat()}.
 #'
 #' @param .f Function to splat.
+#' @param ... Default arguments for \code{.f}.
```
Comment (hadley) [inline R/splat.R]: I'd add: "These will be evaluated once when `splat()` is called".

**Comment by lionelgit** [comment]
Thanks for your reviews Hadley.

One last thing, do you think the point above would be a good idea? That is, making `splat()` behave as a functional when it's given a list, or a factory function when it's given a function. I think it makes sense and would result in prettier pipes.

**Comment by hadley** [comment]
Again, I think that's like currying, which I don't really like for R. Functions are easier to understand when they always return the same type of thing.

**Comment by lionelgit** [comment]
In this case I thought of it more as S3 polymorphism than currying (splatting a list vs splatting a function).

**Comment by hadley** [comment]
Hmmmm, could you share a draft implementation?

**Comment by lionelgit** [comment]
Something like

```{r}
splat <- function(x, ...) {
  UseMethod("splat")
}

#' Splat an object as arguments to a function
#'
#' This calls \code{.f} with the elements of \code{.x} as
#' arguments. This is basically a pipable version of \code{do.call}.
splat.default <- function(.x, .f, ...) {
  defaults <- list(...)
  do.call(".f", c(.x, defaults))
}

#' Splat a function
#'
#' This modifies the function \code{.f} to make it take a list of
#' arguments as argument.
splat.function <- function(.f, ...) {
  force(.f)
  defaults <- list(...)
  function(args = list()) {
    do.call(".f", c(args, defaults))
  }
}
```

```{r}
x <- list(c(1:100, NA, 1000))
splatted_mean <- splat(mean, na.rm = TRUE, trim = 0.9)

x %>% splat(mean, na.rm = TRUE, trim = 0.9)
x %>% splat(mean, na.rm = TRUE, trim = 0.9)()
x %>% splatted_mean()
```

**Comment by lionelgit** [comment]
Not sure if we should use `enlist()` instead of `c()` in the `do.call()` calls.

With `enlist()`,
```
c(1:100, NA, 1000) %>% splat(mean)
```
would work as expected.

But:
```
c("01", "01", "2001") %>% splat(paste, sep = "-")
## [1] "01"   "01"   "2001"
```
```
list("01", "01", "2001") %>% splat(paste, sep = "-")
## [1] "01-01-2001"
```
With `c()` they would both work.

**Comment by hadley** [comment]
That looks ok, but the arguments of the methods must match the generic, so they'd all need to use `.x` as the first argument. Documenting that is going to get nasty. How about `splat()` and `splat_map()` (or `map_splat()`?)

**Comment by lionelgit** [comment]
`map_splat()` would work.

We could also `splat()` a list and `smash()` a function?

**Comment by hadley** [comment]
I love the idea of a function called `smash()`, but I suspect it would be more confusing.

**Comment by lionelgit** [comment]
the version for lists could simply be called `do_call()`
Or `map_call()`

**Comment by lionelgit** [comment]
```
list %>%
  map_call("fun")
```
works pretty well I think

**Comment by hadley** [comment]
Oh I like that

**Comment by lionelgit** [comment]
Can I rename `splat()` to `smash()`?

The idea of `splat`ting arguments to a function suggests that `splat()` takes a list of arguments, not a function.

And this way we'll get to smash functions in R anytime these damn things won't take a list :-)

**Comment by hadley** [comment]
Do it!

**Comment by lionelgit** [comment]
`check()` complains because `map_d()` is still used in examples in some files of this branch (this is fixed in master) and because it does not find `enlist()`.

So after #36 is merged this is safe to merge.

File: R/smash.R [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+#' While `smash()` wraps a function in \code{\link{do.call}()},
+#' `map_call()` is directly equivalent to \code{do.call()} except that
+#' it takes a list as first argument instead of a function. This makes
+#' `map_call()` pipable.
+#' @param .x A list or a vector. Vectors are automatically wrapped in
+#'   a list.
+#' @param .f A function or the name of a function to call with the
+#'   elements of \code{.x} as arguments.
+#' @param ... Additional arguments passed on to \code{.f}.
+#' @seealso smash do.call
```
Comment (hadley) [inline R/smash.R]: These need to be `\code{\link{smash}}` so they turn into links in the docs

**Comment by hadley** [comment]
Looks good - just one tiny niggle left in the docs. 

You can make the travis failure go away by merging master back on to this branch (but it's not a big deal either way)

**Comment by lionelgit** [comment]
I encapsulated the dots in a list so that the result of `c(.x, list(...))` is always a list. This allows `map_call()` to work with atomic vectors, e.g.:
```
c("a", "b") %>% map_call("paste")
```

File: R/smash.R [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+#' @seealso \code{\link{smash}()} and \code{\link{do.call}()}
+#' @export
+#' @examples
+#' # We map a list of strings to paste(), with sep = "-" and the
+#' # string "2001" as additional arguments
+#' list("01", "01") %>%
+#'   map(~ sub("^01", "10", .)) %>%
+#'   map_call("paste", "2001", sep = "-")
+map_call <- function(.x, .f, ...) {
+  do.call(.f, c(.x, list(...)))
```
Comment (hadley) [inline R/smash.R]: `quote(.x)` would be slightly better here

**Comment by hadley** [comment]
One last tiny nitpick - otherwise looks great!

**Comment by lionelgit** [comment]
thanks, here you go :-)

**Comment by lionelgit** [comment]
hmm with the quoted `.x`, `map_call()` does not work with atomic vectors anymore.

**Comment by lionelgit** [comment]
Actually it does not work with lists either. At least not in all situtations:
```
x <- list("a", "b")

do.call("paste", list(quote(.x)))
## [1] "a" "b"

do.call("paste", .x)
## [1] "a b"
```

**Comment by lionelgit** [comment]
with `fnu <- function(...) match.call()` :
```
do.call("fnu", .x)
## fnu("a", "b")
do.call("fnu", list(quote(.x)))
## fnu(.x)
```

**Comment by lionelgit** [comment]
And I just realised that it makes sense to also allow dots in the closure returned by `smash()`

**Comment by hadley** [comment]
Oops, yes, that makes sense.

**Pull request merged**